### PR TITLE
Add support for custom player models

### PR DESCRIPTION
**Pull request by cjohnson57**
Adds support for selecting (on local) or uploading via filepicker (on web) a .zobj definition for a custom player model. Unlike ML64, these models are subject to size and poly count restrictions present in the normal game. There are two types of models, which will be referred to as zzconvert and zzplayas models.

### zzconvert models 
are models which are output from Blender with a custom extension. They only contain original data created by the modeler. Thus, they can be freely shared through the Discord or other means without worries of copyright. In zzconvert models, the LUT is empty. (LUT = LookUp Table, a table in the model that points to the displaylists for all pieces of it.)

### zzplayas models
are one step further in the process, having gone through a process to add all pieces missing from the zzconvert model into it, which is copyrighted content. For example, if the modeler did not create a custom hookshot, zzplayas will add the hookshot from a bank file into the model.

Since zzplayas uses a bank file which is basically just the vanilla Link model, and thus copyrighted content, the code here recreates the functionality of zzplayas to some extent, but it pulls the vanilla data from the user's rom so there should be no copyright concerns.

On local, models can be placed in data/Models/Adult or data/Models/Child and a combobox will fill with the .zobj files in those folders, allowing you to select whichever model you want to use, or just choosing Default for the default player model. If you have more than 1 custom model, there is also a Random option to randomly choose among your custom models. On web, there are instead filepickers where you select the path to your model file to be loaded in. @TreZc0 from my understanding, you and/or db0 will need to add some support for the filepicker to the website once this is merged.

Once the model is loaded from the file, it first checks if it's a playas model (by looking for the string "MODLOADER64" at the beginning of the LUT). If it is, then it needs no further processing and is slapped into the player model portion of the ROM and all the bits are repointed. If it isn't, then the model's footer (created by zzconvert) is checked to see which pieces are present in the model and which are missing and need to be loaded from vanilla. After loading these pieces, the LUT is created by writing pointers to individual pieces as well as writing some constants. It also creates the hierarchy pointer, which is what controls how the different pieces are loaded in the game and relate to each other.

When loading the vanilla data, the displaylist is not enough. It will also contain pointers to texture, matrix, and vertex data, so those will need to be loaded as well, and then repointed in the displaylist so it knows where to look for this data. The vanilla data basically creates a 2nd zobj with just the missing pieces, which is then slapped onto the end of the user-provided zobj and then finally loaded into the ROM.

Certain indices of some pieces are skipped to remove redundancy or unnecessary vanilla data. For example, most vanilla items have Link's fist attached to them, but as zzconvert's hierarchy simply uses the default empty fist and adds the item into it, this is unnecessary- All it does is stick Link's fist inside of whatever the given model's fist is. So the bytes of an item's displaylist that contain the fist are skipped so that only the item remains.

Note that the code here doesn't change the skeleton at all, meaning hit and hurtboxes will all be where they are normally, so a model should offer no competitive advantage (other than if someone makes a model that gives them better visual cues or something I guess.)

This has been tested with a few different zzconvert models so far, as well as on Everdrive.

I tried to comment everything very in-depth since this code is complex and not really part of anyone's existing skillset at OoTR, but feel free to ask here or @ me in the discord if there's something confusing or just needing more explanation.

-----------

Ginormous thanks to @Maplesstar for making models to use, helping me figure out issues, and general emotional support. Also since she will be the one making the guides for model creation so that this feature is easier to use.

Gigantic thanks to @denoflionsx for helping point me in the right direction, understand the ML64 model code, helping sleuth issues, and more- Same goes for @JerryWester, @Neirn, and other helpful folks over at ML64.

Note: These screenshots were all captured from Project64, and these are all zzconvert models.

GLideN64_THE_LEGEND_OF_ZELDA_002

GLideN64_THE_LEGEND_OF_ZELDA_010

GLideN64_THE_LEGEND_OF_ZELDA_005

GLideN64_THE_LEGEND_OF_ZELDA_008

GLideN64_THE_LEGEND_OF_ZELDA_007

### CODE REVIEW

**Comment by denoflionsx** [comment]
I ensured while guiding Shadow through the initial stages of this process that all existing ML64 models that meet the N64 specifications will work. Users of OotR are free to use our [repository of models](https://github.com/hylian-modding/Z64-CustomPlayerModels) to obtain any model available to ML users in raw zobj form.

**Comment by flagrama** [comment]
Has this been tested at all on Virtual Console and Everdrive yet?

**Comment by Maplesstar** [comment]
No, as of right now tested on PJ64 and Retroarch. I intend to do VC tests soon if no one else does first, but everdrive we need someone else.

**Comment by flagrama** [comment]
I did some quick testing with Everdrive. Nothing extensive, but confirmed nothing crashed immediately and went around various areas as child and adult custom models. Did run into a graphical glitch where the Master Sword disappeared when pulling it from the pedestal, but the game didn't crash so seems like it's just a cosmetic issue. Aside from that, someone else should probably do an in-depth test with Everdrive, but it at least runs and so far doesn't seem different than playing with the vanilla model.

**Comment by cjohnson57** [comment]
Confirmed the MS-pulling issue isn't everdrive exclusive, will probably have a fix up tomorrow

**Comment by malonrose** [comment]
Amazing work everyone!
Super excited to see this implemented 👍

**Comment by cjohnson57** [comment]
The code will now check the model's skeleton. If it's very close to the vanilla skeleton, it'll simply be corrected to be vanilla. If not, it draws a message in the pause screen that the model does not have a vanilla skeleton and is thus illegal for racing.

File: ASM/c/model_text.c [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+void draw_illegal_model_text(z64_disp_buf_t *db) {
+
+    // Only draw when paused
+    if (!(illegal_model && z64_game.pause_ctxt.state == 6)) {
+        return;
+    }
+    
+    // Setup draw location
+    int str_len = 36;
+    int total_w = str_len * font_sprite.tile_w + triforce_sprite.tile_w;
```
Comment (fenhl) [inline ASM/c/model_text.c:14]:
```suggestion
    int total_w = str_len * font_sprite.tile_w;
```

Comment (cjohnson57) [reply ASM/c/model_text.c]: Oops, you can tell where I copied this code from, lol

**Comment by cjohnson57** [comment]
Here's what the non-vanilla skeleton message looks like in game:

image

A bit close to the margins unfortunately, but since I can't control font size and don't want it covering any part of the menu or the triforce display, I think this is the best we can do.

**Comment by cjohnson57** [comment]
Using patch files should now work, except for the case where a patch file contains a model and you attempt to override it with a vanilla model. I can't think of any good way to do this currently as all the pointers are overridden to use their ml64 manifest (LUT) values in the patch file, and I'd just need to somehow set them all back to their vanilla values.

**Comment by cjohnson57** [comment]
Ok, all issues with patch files should now be fixed.

**Comment by cjohnson57** [comment]
Note that while this PR has been merged, we still need to find a solution for adding the race-illegal skeleton message to the ROM from the webpatcher.

**Pull request merged**